src/sacn/merge_receiver.c

```diff
@@ -632,123 +635,373 @@ void universe_data(sacn_receiver_t handle, const EtcPalSockAddr* source_addr, co
+    }
+    else if ((header->start_code == 0xDD) && use_pap)
+    {
+      error_status = sacn_dmx_merger_update_paps(merger_handle, source_id, pdata, header->slot_count);
+      new_merge_occurred = true;
+    }
   }
 
   // Notify if needed.
   MergeReceiverMergedDataNotification merged_data_notification = MERGE_RECV_MERGED_DATA_DEFAULT_INIT;
   MergeReceiverNonDmxNotification non_dmx_notification = MERGE_RECV_NON_DMX_DEFAULT_INIT;
 
-  if (sacn_lock())
+  if (error_status == kEtcPalErrOk)
   {
-    SacnMergeReceiver* merge_receiver = NULL;
-    if ((lookup_merge_receiver((sacn_merge_receiver_t)handle, &merge_receiver, NULL) == kEtcPalErrOk))
+    if (sacn_lock())
     {
-      if (new_merge && !is_sampling && (merge_receiver->num_pending_sources == 0))
+      SacnMergeReceiver* merge_receiver = NULL;
+      error_status = lookup_merge_receiver((sacn_merge_receiver_t)handle, &merge_receiver, NULL);
+      if (error_status == kEtcPalErrOk)
       {
-        merged_data_notification.callback = merge_receiver->callbacks.universe_data;
-        merged_data_notification.handle = (sacn_merge_receiver_t)handle;
-        merged_data_notification.universe = header->universe_id;
-        memcpy(merged_data_notification.slots, merge_receiver->slots, DMX_ADDRESS_COUNT);
-        memcpy(merged_data_notification.slot_owners, merge_receiver->slot_owners,
-               DMX_ADDRESS_COUNT * sizeof(sacn_source_id_t));
-        merged_data_notification.context = merge_receiver->callbacks.callback_context;
+        if (new_merge_occurred && !merge_receiver->sampling && (merge_receiver->num_pending_sources == 0))
+        {
+          merged_data_notification.callback = merge_receiver->callbacks.universe_data;
+          merged_data_notification.handle = (sacn_merge_receiver_t)handle;
+          merged_data_notification.universe = header->universe_id;
+          memcpy(merged_data_notification.slots, merge_receiver->slots, DMX_ADDRESS_COUNT);
+          memcpy(merged_data_notification.slot_owners, merge_receiver->slot_owners,
+                 DMX_ADDRESS_COUNT * sizeof(sacn_source_id_t));
+          merged_data_notification.context = merge_receiver->callbacks.callback_context;
+        }
+
+        if ((header->start_code != 0x00) && (header->start_code != 0xDD))
+        {
+          non_dmx_notification.callback = merge_receiver->callbacks.universe_non_dmx;
+          non_dmx_notification.handle = (sacn_merge_receiver_t)handle;
+          non_dmx_notification.universe = header->universe_id;
+          non_dmx_notification.source_addr = source_addr;
+          non_dmx_notification.header = header;
+          non_dmx_notification.pdata = pdata;
+          non_dmx_notification.context = merge_receiver->callbacks.callback_context;
+        }
       }
 
-      if ((header->start_code != 0x00) && (header->start_code != 0xDD))
+      sacn_unlock();
+    }
+    else
+    {
+      error_status = kEtcPalErrSys;
+    }
+  }
+
+  if (error_status == kEtcPalErrOk)
```

Comment:
You technically don't need this check, since the callbacks aren't initialized in the error case.